x-pack/metricbeat/module/aws/awshealth/awshealth.go

```diff
@@ -0,0 +1,437 @@
+// of an error set the Error field of mb.Event or simply call report.Error().
+func (m *MetricSet) Fetch(ctx context.Context, report mb.ReporterV2) error {
+	ctx, cancel := context.WithCancel(ctx)
+	defer cancel()
+
+	var config aws.Config
+	err := m.Module().UnpackConfig(&config)
+	if err != nil {
+		return err
+	}
+
+	awsConfig := m.MetricSet.AwsConfig.Copy()
+
+	health_client := health.NewFromConfig(awsConfig, func(o *health.Options) {
+		if config.AWSConfig.FIPSEnabled {
+			o.EndpointOptions.UseFIPSEndpoint = awssdk.FIPSEndpointStateEnabled
+		}
+	})
+
+	events := m.getEventsSummary(ctx, health_client)
+	for _, event := range events {
+		report.Event(event)
+	}
+
+	return nil
+}
+
+// getEventsSummary retrieves a summary of AWS Health events which are upcoming
+// or open. It uses the DescribeEvents API to get a list of events. Each event is
+// identified by a Event ARN. The function returns a slice of mb.Event structs
+// containing the summarized event info.
+func (m *MetricSet) getEventsSummary(
+	ctx context.Context,
+	awsHealth *health.Client,
+) []mb.Event {
+	var events []mb.Event
+	eventFilter := types.EventFilter{
+		EventStatusCodes: []types.EventStatusCode{
+			types.EventStatusCodeUpcoming,
+			types.EventStatusCodeOpen,
+		},
+	}
+
+	var (
+		nextTokenString string
+		eventOutput     *health.DescribeEventsOutput
+		err             error
+	)
+	errCh := make(chan error, maxResults)
+
+	// Create buffered channel to receive event with event description and affected entity details
+	c := make(chan HealthDetails, maxResults)
+
+	for {
+		// When invoking the DescribeEvents for the first time, there must not exist any NextToken.
+		// Upon calling the DescribeEvents API, the next token will be returned if there are additional records available for querying.
+		// If there are no more records to fetch, the next token will be empty.
+		if nextTokenString == "" {
+			eventOutput, err = awsHealth.DescribeEvents(ctx,
+				&health.DescribeEventsInput{
+					Filter:     &eventFilter,
+					MaxResults: &maxResults,
+				},
+			)
+		} else {
+			eventOutput, err = awsHealth.DescribeEvents(ctx,
+				&health.DescribeEventsInput{
+					Filter:     &eventFilter,
+					MaxResults: &maxResults,
+					NextToken:  &nextTokenString,
+				},
+			)
+		}
+		if err != nil {
+			err = fmt.Errorf("[AWS Health] DescribeEvents failed with : %w", err)
+			m.Logger().Error(err.Error())
+			return nil
+		}
+		ets := eventOutput.Events
+		select {
+		case <-ctx.Done():
+			// Context cancelled, handle graceful termination
+			close(c)
+			return nil
+		default:
+			// Context not cancelled, proceed with the function
+		}
+
+		for i := range ets {
+			m.Logger().Debugf("[AWS Health] [Fetch DescribeEventDetails] Event ARN : %s", getValueOrDefault(ets[i].Arn, ""))
+			go func(et types.Event) {
+				err := m.getDescribeEventDetails(ctx, awsHealth, et, c)
+				if err != nil {
+					errCh <- err
+				}
+			}(ets[i])
+		}
```

Comment:
```suggestion
		for _, event := range ets {
			m.Logger().Debugf("[AWS Health] [Fetch DescribeEventDetails] Event ARN : %s", getValueOrDefault(event.Arn, ""))
			go func(et types.Event) {
				err := m.getDescribeEventDetails(ctx, awsHealth, et, c)
				if err != nil {
					errCh <- err
				}
			}(event)
		}
```

Reply:
If the size of the struct (event) is considerable, then indexing is better rather than element (over-for-range) access as in for range it creates a copy of the element. In this case, ets[i] access is fine.

Reply:
Got your point.

Reply:
resolving this.